flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactCoordinator.java

```diff
@@ -0,0 +1,155 @@
+        inputFiles = new HashMap<>();
+    }
+
+    @Override
+    public void processElement(StreamRecord<CoordinatorInput> element) throws Exception {
+        CoordinatorInput coordinatorInput = element.getValue();
+        if (coordinatorInput instanceof InputFile) {
+            InputFile file = (InputFile) coordinatorInput;
+            // collect the written files
+            inputFiles
+                    .computeIfAbsent(file.getPartition(), k -> new ArrayList<>())
+                    .add(file.getFile());
+        }
+    }
+
+    @Override
+    public void endInput() throws Exception {
+        for (Map.Entry<String, List<Path>> partitionFiles : inputFiles.entrySet()) {
+            compactPartitionFiles(partitionFiles.getKey(), partitionFiles.getValue());
+        }
+    }
+
+    private void compactPartitionFiles(String partition, List<Path> paths) throws IOException {
+        int unitId = 0;
+        final Map<Path, Long> filesSize = getFilesSize(fs, paths);
+        // calculate the average size of these files
+        AverageSize averageSize = getAverageSize(filesSize);
+        if (averageSize.isLessThan(compactAverageSize)) {
+            // we should compact
+            // get the written files corresponding to the partition
+            Function<Path, Long> sizeFunc = filesSize::get;
+            // determine what files should be merged to a file
+            List<List<Path>> compactUnits = BinPacking.pack(paths, sizeFunc, compactTargetSize);
+            for (List<Path> compactUnit : compactUnits) {
+                // emit the compact units containing the files path
+                output.collect(
+                        new StreamRecord<>(new CompactionUnit(unitId++, partition, compactUnit)));
+            }
+        } else {
+            // no need to merge these files, emit each single file to downstream for committing
+            for (Path path : paths) {
+                output.collect(
+                        new StreamRecord<>(
+                                new CompactionUnit(
+                                        unitId++, partition, Collections.singletonList(path))));
+            }
+        }
+    }
+
+    private Map<Path, Long> getFilesSize(FileSystem fs, List<Path> paths) throws IOException {
+        Map<Path, Long> filesStatus = new HashMap<>();
+        for (Path path : paths) {
+            long len = fs.getFileStatus(path).getLen();
+            filesStatus.put(path, len);
+        }
+        return filesStatus;
+    }
+
+    private AverageSize getAverageSize(Map<Path, Long> filesSize) {
+        int numFiles = 0;
+        long totalSz = 0;
+        for (Map.Entry<Path, Long> fileSize : filesSize.entrySet()) {
+            numFiles += 1;
+            totalSz += fileSize.getValue();
+        }
+        return new AverageSize(totalSz, numFiles);
+    }
+
+    private static class AverageSize {
+        private final long totalSz;
+        private final int numFiles;
```

Comment:
average size can be calculated eagerly

Reply:
I don't think it's bad to calculate it later. Also, if the numFiles is zero, we can't calculate the average size.